src/commands/up.rs

```diff
@@ -646,6 +667,87 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+// Scrubs the locked app to only contain the given list of components
+// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, components: &[String]) -> Result<()> {
+    // Create a temporary app to access parsed component and trigger information
+    let tmp_app = spin_app::App::new("tmp", locked_app.clone());
+    let component_trigger_id_map = component_trigger_map(&tmp_app, components)?;
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map
+        .values()
+        .flatten()
+        .collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
+    locked_app.triggers.retain(|t| triggers.contains(&t.id));
+    Ok(())
+}
+
+// Given a list of components and App, returns a map of component ID to trigger ID.
+// Also validates that the components exist in the app and have valid service chaining.
+fn component_trigger_map(
+    app: &spin_app::App,
+    components: &[String],
+) -> Result<ComponentTriggerMap> {
+    let component_triggers: Vec<(AppComponent<'_>, String)> = app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) => {
+                if components.contains(&comp.id().to_string()) {
+                    Some((comp, t.id().to_string()))
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        })
+        .collect();
+
+    // Allow only service chaining among retained components
+    // All wildcard service chaining is disallowed
+    for (c, _) in &component_triggers {
+        let allowed_hosts = allowed_hosts(c)?;
+        allowed_hosts.iter().try_for_each(|host| {
+            let uri = host.parse::<http::Uri>().unwrap();
```

Comment:
I think you might have a problem here with templated URIs?  I believe they're resolved during trigger startup, unless the factors work has changed that.  Might merit a case in the unit tests.

Reply:
We can't (currently) resolve templates here, so I guess there are two questions:

- Will this code panic if it tries to resolve a template? (seems likely to me)

- What _should_ happen with templates here? I would probably suggest doing nothing. Using templates for service chaining seems like an uncommon scenario and the consequence here is pretty low: runtime error instead of this nicer startup validation.

Reply:
Yes, the current code will panic on an unsubstituted template.

I agree that ignoring the error is likely best.  Service chaining via a templated URL is one of those hazy features.  It _happens_ to work in the current CLI implementation, and I don't believe we complain if someone does it, but there are no guarantees around it.  So it seems reasonable for a failed URL parse to be interpreted as "it's not a service chaining URL" which means it is of no further interest.

If we wanted more belt and braces we could validate during trigger load (after template substitution) that all non-wildcard service chaining URLs pointed to components that exist in the app.  Which might not be a bad plan anyway, and can be done outside the scope of this PR.

Reply:
Good call out here. I agree that we should ignore URLs that cannot be parsed for now and +1 to validating non-wildcard service chaining URLs are to existing components.